chore: run typescript build as part of typecheck, not normal build

SimenB opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

That way type errors will not block unit tests from running

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

## Test plan

Green CI

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

SimenB commented on package.json:
```diff
@@ -84,14 +84,14 @@
     "lint:md:ci": "prettylint '**/*.{md,yml,yaml}' --ignore-path .gitignore",
     "postinstall": "opencollective postinstall && yarn build",
-    "publish": "yarn build-clean && yarn build && lerna publish --silent",
+    "publish": "yarn build-clean && yarn typecheck && yarn build && lerna publish --silent",
```
@rubennorte is this fine? We need to ensure the typescript build is ran so the type definition files are generated.

If you _don't_ use this script, anyway else we can ensure it runs before publishing?

jeysal replied:
Possibly cleaner alternative would be doing a tsc `noEmit` for `typecheck` and a tsc `emitDeclarationOnly`  for `build`

SimenB replied:
you cannot use `noEmit` with references. And even if it did, that doesn't make sense to me. Why do the same work twice? You cannot emit without typechecking anyways, so it's just more work for what I'd say is zero gain.

Keeping the typecheck out of `build` means we can run tests on CI even if types fail. Different signals

jeysal replied:
Okay, I get why that wouldn't make sense for us. Works better in projects where a compile step is only needed for production, not for development/testing.

rubennorte replied:
@SimenB if that command generates definitions it should be part of the `build` script. We don't always use the `publish` command but we always run the build.

SimenB replied:
Alright, that seems fair :) Will slow down CI a bit since we run `build` as `postinstall`, but it's not too bad

SimenB commented on scripts/typeScriptBuild.js:
```diff
@@ -0,0 +1,24 @@
+  fs.existsSync(path.resolve(p, 'tsconfig.json'))
+);
+
+execa.sync('tsc', ['-b', ...packageWithTs, ...process.argv.slice(2)], {
```
Makes `node scripts/typeScriptBuild.js --watch` work

thymikee commented on package.json:
```diff
@@ -84,14 +84,14 @@
     "test-leak": "yarn jest -i --detectLeaks jest-mock jest-diff jest-repl",
     "test": "yarn typecheck && yarn lint && yarn jest",
-    "typecheck": "flow check --include-warnings",
+    "typecheck": "node scripts/typeScriptBuild && flow check --include-warnings",
```
How about naming it `buildTS`

thymikee commented on scripts/typeScriptBuild.js:
```diff
@@ -0,0 +1,24 @@
+
+const packages = getPackages();
+
+const packageWithTs = packages.filter(p =>
```
Typo: packagesWithTs

SimenB commented:
Ok, ended up adding `build:ts` which builds type definitions. Also added `prebuild: build:ts`, so it triggers automatically before `build`. Also added `watch:ts` which runs `tsc` in watch mode
